src/main/scala/com/fulcrumgenomics/umi/FilterConsensusReads.scala

```diff
@@ -0,0 +1,217 @@
+    val header   = in.getFileHeader.clone()
+    header.setSortOrder(SortOrder.coordinate)
+    val sorter    = Bams.sortingCollection(SortOrder.coordinate, header)
+    val out       = new SAMFileWriterFactory().setCreateIndex(true).makeWriter(header, true, output.toFile, null)
+    val progress1 = new ProgressLogger(logger, verb="Filtered and masked")
+
+    // Go through the reads by template and do the filtering
+    Bams.templateIterator(in).foreach { template =>
+      val r1 = template.r1.getOrElse(throw new IllegalStateException(s"${template.name} had no R1."))
+      if (r1.getReadPairedFlag) require(template.r2.isDefined, s"Paired read missing R2: ${template.name}")
+      totalReads += template.readCount
+
+      val r1Result = filterRecord(r1)
+      val r2Result = template.r2.map(filterRecord).getOrElse(FilterResult(keepRead=true, 0))
+
+      if (r1Result.keepRead && r2Result.keepRead) {
+        totalBases  += r1.getReadLength + template.r2.map(_.getReadLength).getOrElse(0)
+        maskedBases += r1Result.maskedBases + r2Result.maskedBases
+        sorter.add(r1)
+        progress1.record(r1)
+        template.r2.foreach { r => sorter.add(r); progress1.record(r) }
+
+        template.allSupplementaryAndSecondary.foreach { r =>
+          val result = filterRecord(r)
+          if (result.keepRead) {
+            totalBases += r.getReadLength
+            maskedBases += result.maskedBases
+            sorter.add(r)
+            progress1.record(r)
+          }
+        }
+      }
+    }
+
+    // Then iterate the reads in coordinate order and re-calculate key tags
+    logger.info("Filtering complete; fixing tags and writing coordinate sorted reads.")
+    val progress2 = new ProgressLogger(logger, verb="Wrote")
+    val walker = new ReferenceSequenceFileWalker(ref.toFile)
+    sorter.doneAdding()
+    sorter.foreach { rec =>
+      Bams.regenerateNmUqMdTags(rec, walker)
+      out.addAlignment(rec)
+    }
+
+    in.safelyClose()
+    out.close()
+    logger.info(f"Output ${progress1.getCount}%,d of ${totalReads}%,d consensus reads.")
+    logger.info(f"Masked ${maskedBases}%,d of ${totalBases}%,d bases in retained consensus reads.")
+  }
+
+  /**
+    * Perform filtering on an individual read.
+    *
+    * @param  rec the read to filter
+    * @return a tuple of (Boolean, Int) where the first value is true if the read should be kept
+    *         and false if it should be filtered out, and the second value is the count of bases
+    *         in the read that were masked to Ns as a result of per-base filtering
+    */
+  private[umi] def filterRecord(rec: SAMRecord): FilterResult = {
+    val maxDepth = rec.getIntegerAttribute(ConsensusTags.PerRead.RawReadCount)
+    val errRate = rec.getFloatAttribute(ConsensusTags.PerRead.RawReadErrorRate)
+    if (maxDepth == null || errRate == null) fail(s"Read ${rec.getReadName} does not appear to have consensus calling tags present.")
```

Comment:
We currently have no way to create that situation, so I think that would be un-needed complexity.